### PR TITLE
[CALCITE-3823] Do not use String.replaceAll

### PR DESCRIPTION
**Pull request by julianhyde**
String.replaceAll uses regex, which is inefficient, and may not be
what we want. For strings, use either String.replace; for regex use
Pattern.compile().matcher().replaceAll(), being sure to store the
pattern for future use.

Add entries to forbidden-apis/signatures.txt to prevent people
using String.replaceAll in future.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 1]
```diff
@@ -346,14 +346,7 @@ public RelDataTypeSystem getTypeSystem() {
...
   public String quoteIdentifier(String val) {
-    if (identifierQuoteString == null) {
-      return val; // quoting is not supported
-    }
-    String val2 =
-        val.replaceAll(
-            identifierEndQuoteString,
-            identifierEscapedQuote);
-    return identifierQuoteString + val2 + identifierEndQuoteString;
+    return quoteIdentifier(new StringBuilder(), val).toString();
```
Comment (amaliujia) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: just a side note: jdk has some sort of string concatenation optimizations after JDK9. In the future after JDK usage is upgrade, there might be less need to do such rewrite.

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/sql/SqlDialect.java]: Fair point.

I did it mainly to remove code (unify the two `quoteIdentifier` methods, and remove the override for `BigQueryDialect`).

**Comment by hsyuan** [comment]
+1

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 2]
```diff
@@ -1694,16 +1696,18 @@ public static String toString(double x) {
...
+    return PATTERN_0_STAR_E.matcher(s).replaceAll("E").replace("E+", "E");
   }
 
   /** CAST(DECIMAL AS VARCHAR). */
   public static String toString(BigDecimal x) {
     final String s = x.toString();
-    if (s.startsWith("0")) {
+    if (s.equals("0")) {
+      return s;
+    } else if (s.startsWith("0.")) {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: Is this locale-dependent?
The use of dot here does not look right.

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: That may or may not be a bug. If there's a bug (e.g. you can reproduce in SQL) please log it. I refactored the existing code, so I don't think I made anything worse.

Comment (julianhyde) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: The `Float.toString()`is not locale-dependent - that is, it always uses '.' as decimal separator - and I think that's what we want for `CAST(REAL AS VARCHAR)`.

**Comment by DonnyZone** [comment]
Fixed in https://github.com/apache/calcite/commit/4208d0ba6f2a749692fe64181a1373af07d55db5.

**Pull request closed without merging**